### ISSUE TITLE
cgroup_get_cgroup() broken for cgroupv2?

### ISSUE THREAD

**Issue by adriaan42** [libcgroup/libcgroup#272]
I'm running libcgroup v3.0.0 on Linux 6.1:
```sh
# uname -a
Linux box 6.1.0-0-rt-amd64 #1 SMP PREEMPT_RT Debian 6.1~rc3-1~exp1 (2022-11-02) x86_64 GNU/Linux
# mount | grep cgroup
cgroup2 on /sys/fs/cgroup type cgroup2 (rw,nosuid,nodev,noexec,relatime,nsdelegate,memory_recursiveprot)
```

With the following reproducer:
```c
#include <stdlib.h>
#include <unistd.h>
#include <libcgroup.h>
#include "libcgroup-internal.h"

struct cgroup* get_or_create_cgroup(char *name)
{
    struct cgroup *cg = cgroup_new_cgroup(name);
    int ret = cgroup_get_cgroup(cg);
    if (ret == ECGROUPNOTEXIST) {
        struct cgroup_controller *cgc = cgroup_add_controller(cg, "cpuset");
        int ret2 = cgroup_create_cgroup(cg, 1);
        if (ret2) {
            printf("failed to create %s: %s\n", name, cgroup_strerror(ret2));
            exit(1);
        }
        printf("created %s\n", name);
    } else if (ret) {
        printf("failed to get %s: %s\n", name, cgroup_strerror(ret));
        exit(1);
    } else {
        printf("found existing %s\n", name);
#ifdef WORKAROUND
        cgroup_free(&cg);
        cg = cgroup_new_cgroup(name);
#endif
    }
    printf("  controllers:");
    for (int i = 0; i < CG_CONTROLLER_MAX; i++) {
        if (!cg->controller[i])
            break;
        printf(" %s", cg->controller[i]->name);
    }
    printf("\n");
    return cg;
}

int main(int, char*[]) {
    int ret = cgroup_init();
    if (ret) {
        printf("failed to initialize: %s\n", cgroup_strerror(ret));
        exit(1);
    }

    get_or_create_cgroup("my.slice");
    struct cgroup *cg = get_or_create_cgroup("my.slice/sub.slice");
    ret = cgroup_attach_task(cg);
    if (ret) {
        printf("failed to attach thread: %s\n", cgroup_strerror(ret));
        exit(1);
    }
    printf("success!\n");
}
```

And I'm seeing:
```sh
# gcc -o cgtest cgtest.c -lcgroup
# rmdir /sys/fs/cgroup/my.slice/sub.slice/ /sys/fs/cgroup/my.slice/
# ./cgtest
created my.slice
  controllers: cpuset
created my.slice/sub.slice
  controllers: cpuset
success!
# ./cgtest
found existing my.slice
  controllers: cpuset cpu io memory hugetlb pids rdma misc
found existing my.slice/sub.slice
  controllers: cpuset cpu io memory hugetlb pids rdma misc
failed to attach thread: Cgroup is not mounted
```
It looks to me that
* `cgroup_get_cgroup()` always populates the cgroup struct with all possible controllers, even if they're not in `cgroup.controllers`
* `cgroup_attach_task()` wants to attach the task to all controllers in the cgroup struct, and fails (returns error) if one of them doesn't actually exist. 
* However, if the first controller **does** exist, the task is actually attached, but still an error is returned.

If I use the cgroup right after `cgroup_new_cgroup()` without calling `cgroup_get_cgroup()`, there is no error, and the task is attached to the cgroup.
```sh
# gcc -o cgtest cgtest.c -lcgroup -DWORKAROUND
# rmdir /sys/fs/cgroup/my.slice/sub.slice/ /sys/fs/cgroup/my.slice/
# ./cgtest
created my.slice
  controllers: cpuset
created my.slice/sub.slice
  controllers: cpuset
success!
# ./cgtest
found existing my.slice
  controllers:
found existing my.slice/sub.slice
  controllers:
success!
```

**Comment by kamalesh-babulal** [comment]
@adriaan42 Thank you for the bug report and reproducer.  I am able to reproduce the bug with `main` branch.  Sorry, I would not able to get the earlier and will not be able to until mid-next week.

**Comment by drakenclimber** [comment]
I pushed a fix for this to both the main and release-3.0 branches.  I'll start working on a fix for this in the release-2.0 branch.

Thanks for the bug report, @adriaan42

**Comment by drakenclimber** [comment]
After thinking about this some more, I don't think I want to backport this to the release-2.0 branch.  My reasoning:
* We don't have automated testing around the C APIs in the release-2.0 branch.  Thus this would have to be entirely tested by hand
* It's not a clean backport.  I don't think it would be a hard backport, but since it's not clean there's the chance we could introduce other issues
* I want to do one more release of the release-2.0 branch and then retire it.  If we do introduce other issues, that could cause this branch to continue on
* Ultimately it would be best for users to move to the release-3.0 branch.  It supports systemd, it's better tested, and it's being actively maintained and developed :)
* Finally, no one has complained about it on the release-2.0 branch :/
